kubelet: Introduce PodInfraContainerChanged().

yifan-gu opened the pull request:
@vmarmol @yujuhong 
This new function computes in ahead whether we need to restart the pod
infra container.

Splitted from https://github.com/GoogleCloudPlatform/kubernetes/pull/6169 to enable smoother review.

Tests are broken, fixing...

/cc @jonboulle

vmarmol commented on pkg/kubelet/dockertools/docker.go:
```diff
@@ -40,8 +40,9 @@ import (
-	DockerPrefix          = "docker://"
+	PodInfraContainerName  = leaky.PodInfraContainerName
+	DockerPrefix           = "docker://"
+	PodInfraContainerImage = "gcr.io/google_containers/pause:0.8.0"
```
There is a flag for this, we should use it instead of this (the old version was unused I believe).

yifan-gu replied:
Gotcha

yifan-gu replied:
@vmarmol I have checked about this. The flag is stored in kubelet, I think we should move it to the DockManager?

vmarmol replied:
SGTM, the PR was merged :)

vmarmol commented on pkg/kubelet/kubelet.go:
```diff
@@ -1116,8 +1112,15 @@ func (kl *Kubelet) computePodContainerChanges(pod *api.Pod, runningPod kubeconta
+			return podContainerChangesSpec{}, err
+		}
+		if !changed {
+			glog.V(4).Infof("Infra pod is good, keeps it %q", podFullName)
```
typo: "keep"

yifan-gu replied:
thx!

vmarmol commented on pkg/kubelet/kubelet.go:
```diff
@@ -1116,8 +1112,15 @@ func (kl *Kubelet) computePodContainerChanges(pod *api.Pod, runningPod kubeconta
 		glog.V(4).Infof("Found infra pod for %q", podFullName)
-		podInfraContainerID = dockertools.DockerID(podInfraContainer.ID)
-		containersToKeep[podInfraContainerID] = -1
+		changed, err := kl.containerManager.PodInfraContainerChanged(pod, podInfraContainer)
```
In this case we need to create the podInfra container no?

yifan-gu replied:
Good catch. Refactor the logic a little bit.

vmarmol commented on pkg/kubelet/kubelet.go:
```diff
@@ -1163,21 +1166,10 @@ func (kl *Kubelet) computePodContainerChanges(pod *api.Pod, runningPod kubeconta
 						continue
 					}
 					glog.Infof("pod %q container %q is unhealthy (probe result: %v). Container will be killed and re-created.", podFullName, container.Name, result)
-					containersToStart[index] = empty{}
```
Why are we no longer keeping track of containers to start?

yifan-gu replied:
It's moved below

vmarmol replied:
Ah yes I see that now, thanks!

vmarmol commented on pkg/kubelet/dockertools/manager.go:
```diff
@@ -452,3 +452,40 @@ func makeCapabilites(capAdd []api.CapabilityType, capDrop []api.CapabilityType)
+
+// PodInfraContainer returns true if the pod infra container has changed.
+func (self *DockerManager) PodInfraContainerChanged(pod *api.Pod, podInfraContainer *kubecontainer.Container) (bool, error) {
+	// Use host networking if specified and allowed.
```
nit: comment seems out of place.

vmarmol commented on pkg/kubelet/dockertools/manager.go:
```diff
@@ -452,3 +452,40 @@ func makeCapabilites(capAdd []api.CapabilityType, capDrop []api.CapabilityType)
+		Ports: ports,
+	}
+
+	expectedHash := HashContainer(container)
```
Any reason to use a hash instead of a semantic equals?

yifan-gu replied:
@vmarmol I have no strong opinion on this. Just because the hash approach requires less effort and won't require we parsing the result of 'docker inspect'.
Updated with a deep equal version one.

yifan-gu commented:
@vmarmol Hmm, I found by comparing the container directly, it's making the fake docker much complex since I need to construct a valid `docker.Container` in `StartContainer` and `CreateContainer`. What's the arguments for not using hash actually?

vmarmol commented:
@yifan-gu I'm fine going the has route since we use that today, but it just seems out of place when we do equals other places today.

yifan-gu commented:
@vmarmol Could you point some places where we do equals please? I didn't find one, maybe it's differently and better than me.

vmarmol commented:
@yifan-gu we do it for [PodStatus](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubelet/status_manager.go#L77) and for [MirrorPod](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubelet/pod_manager.go#L287)

yifan-gu commented:
@vmarmol I see. Ideally this can be done in the future when we can reconstruct api.Container from the container runtime.

BTW, I fixed all the tests, to make it easier for you to review, I put those in the second commits :)  I am running e2e now to see if this change works right.

pmorie commented on pkg/kubelet/kubelet_test.go:
```diff
@@ -512,7 +546,11 @@ func TestSyncPodsWithTerminationLog(t *testing.T) {
+		"list",                                 // Get pod status.
+		"create", "start", "inspect_container", // Create pod infra container.
+		"create", "start", // Create container.
+		"list", "inspect_container", "inspect_container"}) // Get pod status.
```
This is classy newline use; make the comments justified at the same position?

pmorie replied:
(and everywhere else) @yifan-gu

pmorie commented:
Is `PodInfraContainerChanged` going to become part of the `Container.Runtime` interface eventually?

yifan-gu commented:
@pmorie I don't think so, it's docker specific IMO, rkt doesn't need a pod infra container.

yifan-gu commented:
Oops, got many e2e failures. 
Debugging...

```
Summarizing 14 Failures:

[Fail] Services [It] should correctly serve identically named services in different namespaces on different external IP addresses 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:373

[Fail] kubectl guestbook [It] should create and stop a working application 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/util.go:304

[Fail] Density [It] should allow starting 3 pods per node 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/density.go:126

[Fail] Networking [It] should function for intra-pod communication 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/networking.go:273

[Fail] kubectl update-demo [It] should scale a replication controller 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/util.go:304

[Fail] emptyDir [It] volume on tmpfs should have the correct mode 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/util.go:325

[Fail] Pods [It] should be submitted and removed 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/pods.go:129

[Fail] ReplicationController [It] should serve a basic image on each replica with a private image 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/rc.go:126

[Fail] Shell [It] tests that services.sh passes 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/shell.go:69

[Fail] Services [It] should be able to create a functioning external load balancer 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:282

[Fail] ReplicationController [It] should serve a basic image on each replica with a public image 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/rc.go:95

[Fail] kubectl update-demo [It] should do a rolling update of a replication controller 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/util.go:304

[Fail] Pods [It] should be restarted with a /healthz http liveness probe 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/pods.go:68

[Fail] PD [It] should schedule a pod w/ a RW PD, remove it, then schedule it on another host 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/pd.go:86

Ran 33 of 39 Specs in 1112.625 seconds
FAIL! -- 19 Passed | 14 Failed | 1 Pending | 5 Skipped F0410 10:38:47.441049   28149 driver.go:112] At least one test failed
2015/04/10 10:38:47 Error running Ginkgo tests: exit status 255
exit status 1

```

yifan-gu commented:
Got a log of connection refused error, such as
```
NFO: After 10.313296623s failed to make proxy call to elasticsearch-logging: Get https://130.211.156.16/api/v1beta3/proxy/namespaces/default/services/elasticsearch-logging/_search?q=log%3Aes_logging_5752synthlogger&size=400: dial tcp 130.211.156.16:443: connection refused

```

Investigating

yifan-gu commented:
Seems like the containers in master is keeping restarting... Sounds like a bug.

yifan-gu commented:
Got 2 failures now:

```
Summarizing 2 Failures:

[Fail] Services [It] should be able to create a functioning external load balancer
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:282

[Fail] kubectl guestbook [It] should create and stop a working application
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/util.go:304

Ran 33 of 39 Specs in 909.800 seconds
FAIL! -- 31 Passed | 2 Failed | 1 Pending | 5 Skipped

```

yifan-gu commented:
Fixed, it's because the target pools are not cleared.

```
Ran 33 of 39 Specs in 1216.063 seconds
SUCCESS! -- 33 Passed | 0 Failed | 1 Pending | 5 Skipped Shutting down test cluster in background
```

@vmarmol

Commit:
kubelet: Introduce PodInfraContainerChanged().

This functions computes in ahead whether we need to restart the pod
infra container.

Commit:
kubelet: Introduce PodInfraContainerChanged().

Update tests.

yifan-gu commented:
@vmarmol I think this is ready for another review?

vmarmol commented:
Awesome, I'll take a look today @yifan-gu

yifan-gu commented:
@vmarmol  Thank you! FYI the 1st commit is the change made to source, 2nd is updating tests. I am running e2e again :)

yifan-gu commented:
Have run for twice again, one passed all tests, one got a failure which I think is a known issue #6424
```
Summarizing 1 Failure:

[Fail] ReplicationController [It] should serve a basic image on each replica with a public image
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/rc.go:151

Ran 33 of 39 Specs in 1219.849 seconds
FAIL! -- 32 Passed | 1 Failed | 1 Pending | 5 Skipped F0413 10:42:57.441241    4053 driver.go:115] At least one test failed
```

vmarmol commented:
Outside of that one change (and the needed rebase), this looks good :)

yifan-gu commented:
@vmarmol Rebased and ran e2e again. Looks good:

```
Ran 35 of 39 Specs in 1163.674 seconds
SUCCESS! -- 35 Passed | 0 Failed | 0 Pending | 4 Skipped Shutting down test cluster in background.
Bringing down cluster using provider: gce
```

vmarmol commented:
LGTM, thanks for the diligence @yifan-gu! Will merge once the CIs go green.

yifan-gu commented:
@vmarmol You're very welcome! I think cleaning up the sync logic might have potential to break the e2e, so I am trying to be more careful. I will keep an eye on that if any stuff fails in the future.

pmorie commented:
Clean GCE e2e run?  It's a miracle.

On Mon, Apr 13, 2015 at 9:22 PM, Yifan Gu <notifications@github.com> wrote:

> @vmarmol <https://github.com/vmarmol> You're very welcome! I think
> cleaning up the sync logic might have potential to break the e2e, so I am
> trying to be more careful. I will keep an eye on that if any stuff fails in
> the future.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6608#issuecomment-92552981>
> .
>
